### PR TITLE
Add Sacrum, Enhance SC Seg & Expand DA for Multi-Contrast

### PR DESCRIPTION
**Pull request by yw7**
This pull request introduces several significant improvements to our model training process and dataset handling:
- Incorporate spinal cord (SC) segmentation for the SPIDER dataset
- Add sacrum segmentation for both SPIDER dataset and select images from the whole-spine dataset
- Implement more aggressive DA methods to improve model robustness and extend support for a wider range of image contrasts

### CODE REVIEW

**Comment by NathanMolinier** [comment]
I'm currently following the readme and running the installation and I noticed different things:
- we are not bound to a specific python version
- we are no bound to specific versions of the packages either in the `pyproject.toml`
Maybe it is fine but I'm also thinking that it could lead to unexpected errors happening with the newest versions of the packages. We could also keep not specific versions for the main branch and only fix versions when doing releases.

**Comment by NathanMolinier** [comment]
Regarding the installation of the packages:
- I installed torch using the command `pip3 install torch torchvision torchaudio`
- Then I installed the `totalspineseg` package and the dependencies by running `python -m pip install -e .`

However, I had an version error probably related to the installation of the package `nnunetv2`
Screenshot 2024-06-27 at 11 49 46

To fix this, I recreated a virtual environment and only ran the `totalspineseg` package installation. This time I did not get any errors so I will try to execute the next steps with this new environment. I'm thinking the extra torch installation might not be needed.

**Comment by yw7** [comment]
> I'm currently following the readme and running the installation and I noticed different things:
> 
> * we are not bound to a specific python version
> * we are no bound to specific versions of the packages either in the `pyproject.toml`
>   Maybe it is fine but I'm also thinking that it could lead to unexpected errors happening with the newest versions of the packages. We could also keep not specific versions for the main branch and only fix versions when doing releases.

regarding pytohn there is limitation.
https://github.com/neuropoly/totalspineseg/blob/395e347bc02260b6ae8658f229d60280f6cdd142/pyproject.toml#L4C1-L4C26
But you're right. Specifying the version could prevent errors however leading to ling time reinstalling all the packages.
We can add specific versions or range in pyproject.toml

**Comment by yw7** [comment]
> Regarding the installation of the packages:
> 
> * I installed torch using the command `pip3 install torch torchvision torchaudio`
> * Then I installed the `totalspineseg` package and the dependencies by running `python -m pip install -e .`
> 
> However, I had an version error probably related to the installation of the package `nnunetv2` Screenshot 2024-06-27 at 11 49 46
> 
> To fix this, I recreated a virtual environment and only ran the `totalspineseg` package installation. This time I did not get any errors so I will try to execute the next steps with this new environment. I'm thinking the extra torch installation might not be needed.

The problem is that some cuda need specific version of pytorch so nneed to do it as describe in their website. Beside this is also how it is recommanded in nnUnet:
https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/installation_instructions.md#installation-instructions.

**Comment by NathanMolinier** [comment]
> regarding pytohn there is limitation. https://github.com/neuropoly/totalspineseg/blob/395e347bc02260b6ae8658f229d60280f6cdd142/pyproject.toml#L4C1-L4C26 

Maybe we could precise this constraint when creating the environment at the beginning of the README ?

**Comment by NathanMolinier** [comment]
> The problem is that some cuda need specific version of pytorch so nneed to do it as describe in their website. Beside this is also how it is recommanded in nnUnet: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/installation_instructions.md#installation-instructions.

You are right indeed ! I wonder how we could make this installation more robust

**Comment by NathanMolinier** [comment]
I am getting an error about a missing package called `funseg` when running `bash "$TOTALSPINESEG"/scripts/prepare_nnunet_datasets.sh`

**Comment by NathanMolinier** [comment]
I just ran this command:
```
bash "$TOTALSPINESEG"/scripts/prepare_nnunet_datasets.sh
```
and noticed that the generated folder is taking more than 800 Go. Also, I was not able to run completely the command due to the size limitation on my device. 

So I think we should first warn the user about the total storage needed for the full command and maybe create a flag to limit the total data augmentation ? Also, how can I specify the step for which I want to generate the data ?

**Comment by yw7** [comment]
> * we are not bound to a specific python version
> Maybe we could precise this constraint when creating the environment at the beginning of the README ?

I've update the ERADME.

> * we are no bound to specific versions of the packages either in the `pyproject.toml`
>   Maybe it is fine but I'm also thinking that it could lead to unexpected errors happening with the newest versions of the packages. We could also keep not specific versions for the main branch and only fix versions when doing releases.

I've opened an issue for this:
https://github.com/neuropoly/totalspineseg/issues/36


> and noticed that the generated folder is taking more than 800 Go. Also, I was not able to run completely the command due to the size limitation on my device.

I've update the README to include hardware specifications for training.


> Why is the data used for step 2 a lot bigger than step 1 ?
This is regarding this issue:
https://github.com/neuropoly/totalspineseg/issues/30

> So I think we should first warn the user about the total storage needed for the full command and maybe create a flag to limit the total data augmentation ? Also, how can I specify the step for which I want to generate the data ?
Done! see README.

**Comment by NathanMolinier** [comment]
I try to run the inference using the weights in the [release](https://github.com/neuropoly/totalspineseg/releases/tag/r20240628).

But it did not work, see:
Screenshot 2024-07-10 at 15 11 48

The script is looking for weights named `Dataset101_*_fold_0.zip` whereas the weights that I downloaded are actually called `Dataset101_TotalSegMRI_step1_fold_all.zip` and `Dataset102_TotalSegMRI_step2_fold_all.zip`.

File: README.md [hunk 1]
```diff
@@ -1,95 +1,118 @@
...
-    ```
-    source totalsegmentator-mri/run/train_nnunet.sh
-    ```
+## Inference
 
-### First Model - Inference
-Run the model on a folder containing the images in .nii.gz format (Make sure to train the model or extract the trained `nnUNet_results` into `data/nnUNet/nnUNet_results` befor running):
-```
-source totalsegmentator-mri/run/inference_nnunet.sh INPUT_FOLDER OUTPUT_FOLDER
+1. If you didn't train the model yourself, you should download the model zip file from the release into `$TOTALSPINESEG_DATA/nnUNet/exports` (without extracting, you can run `mkdir -p "$TOTALSPINESEG_DATA"/nnUNet/exports` before).
```
Comment (NathanMolinier) [inline README.md:104]:
```suggestion
1. If you didn't train the model yourself, you should download the model zip file from the release into `$TOTALSPINESEG_DATA/nnUNet/exports` (do not extract the zip files)
> you can run `mkdir -p "$TOTALSPINESEG_DATA"/nnUNet/exports` to create the folder.
```

**Comment by yw7** [comment]
> I try to run the inference using the weights in the [release](https://github.com/neuropoly/totalspineseg/releases/tag/r20240628).
> 
> But it did not work, see: Screenshot 2024-07-10 at 15 11 48
> 
> The script is looking for weights named `Dataset101_*_fold_0.zip` whereas the weights that I downloaded are actually called `Dataset101_TotalSegMRI_step1_fold_all.zip` and `Dataset102_TotalSegMRI_step2_fold_all.zip`.

Did you checkout to r20240628?
Or follow the instructions on the release page?
git clone https://github.com/neuropoly/totalspineseg.git -b r20240628

**Comment by NathanMolinier** [comment]
> @NathanMolinier Did you checkout to r20240628? Or follow the instructions on the release page? git clone https://github.com/neuropoly/totalspineseg.git -b r20240628

No I did not ! I just downloaded the weights from the release actually and stayed on this branch

**Comment by yw7** [comment]
> No I did not ! I just downloaded the weights from the release actually and stayed on this branch

So this could be the problem. Because the tag r20240628 is made to support this specific version of the model wich is not the same as the defauls in the main branch.

**Comment by yw7** [comment]
> So what should I do if I want to follow this README and use weights from the release as mentioned ?

This part of the README is for the version of the model we're currently training. And we will put it in nex release.
This is a good point. Maybe we should just write "follow the instructions on the release page"..

**Comment by NathanMolinier** [comment]
> This is a good point. Maybe we should just write "follow the instructions on the release page"..

Or maybe the inference script could try to find more weights ? And If multiple weights are available we could raise a warning ?

**Comment by yw7** [comment]
> Or maybe the inference script could try to find more weights ? And If multiple weights are available we could raise a warning ?

It is possible. However, there are some differences in labels between this release (which uses C1) and the current version (which uses C2). So, you'll need to check out that branch anyway. Additionally, it's reasonable to check out the release tag before using the release. Besides that, it works the same, so maybe there's no need for changes in the README.

**Comment by NathanMolinier** [comment]
> It is possible. However, there are some differences in labels between this release (which uses C1) and the current version (which uses C2). So, you'll need to check out that branch anyway.

Is it possible to try the weights before merging this PR ?

**Comment by yw7** [comment]
> > It is possible. However, there are some differences in labels between this release (which uses C1) and the current version (which uses C2). So, you'll need to check out that branch anyway.
> 
> 
> 
> Is it possible to try the weights before merging this PR ?
> 
> 

What do you mean?
I can send you the current checkpoint to try if you want.
I'll send you on slack.

**Comment by jcohenadad** [comment]
> I can send you the current checkpoint to try if you want.

a better procedure is to make a new release for each change of the model, and include the weights/model in the release assets-- so you can already create a release draft, upload the weights there-- when/if PR is merged, release can be published alongside the model assets-- no more slack binary exchanges please

**Comment by yw7** [comment]
> > I can send you the current checkpoint to try if you want.
> 
> 
> 
> a better procedure is to make a new release for each change of the model, and include the weights/model in the release assets-- so you can already create a release draft, upload the weights there-- when/if PR is merged, release can be published alongside the model assets-- no more slack binary exchanges please

I've created a release for the previous version. The current version is during training. So I can create a pre-release with current checkpoint (latest not final) and later delete this release.

**Comment by jcohenadad** [comment]
> I've created a release for the previous version. The current version is during training. So I can create a pre-release with current checkpoint (latest not final) and later delete this release.

if the model won't be released, then no, don't open a pre-release for that. Slack it is then...

**Comment by NathanMolinier** [comment]
I ran this command to generate the dataset for step1:
```bash
bash "$TOTALSPINESEG"/scripts/prepare_nnunet_datasets.sh 101 -noaug
```
And got this issue
Screenshot 2024-07-12 at 09 32 41

**Comment by yw7** [comment]
> I ran this command to generate the dataset for step1:
> 
> ```shell
> bash "$TOTALSPINESEG"/scripts/prepare_nnunet_datasets.sh 101 -noaug
> ```
> 
> And got this issue Screenshot 2024-07-12 at 09 35 52

This error is from old coomit that uses `derivatives/labels` but now we're using `derivatives/labels_iso`  so this should not happen. It happened if there are 4D images in the dataset. Anyway, I've pushed a fix that average all 4D images into 3D images if exists.

**Comment by yw7** [comment]
> Regarding the installation of the packages:
> 
> * I installed torch using the command `pip3 install torch torchvision torchaudio`
> * Then I installed the `totalspineseg` package and the dependencies by running `python -m pip install -e .`
> 
> However, I had an version error probably related to the installation of the package `nnunetv2` Screenshot 2024-06-27 at 11 49 46
> 
> To fix this, I recreated a virtual environment and only ran the `totalspineseg` package installation. This time I did not get any errors so I will try to execute the next steps with this new environment. I'm thinking the extra torch installation might not be needed.

I've update the README to better explain the installation process.
See https://github.com/neuropoly/totalspineseg/blob/retrain-with-sc-and-sacrum/README.md#installation

File: README.md [hunk 2]
```diff
@@ -1,95 +1,151 @@
...
+   ```
+
+1. Prepare datasets in nnUNetv2 structure into `$TOTALSPINESEG_DATA/nnUnet`:
+   ```bash
+   bash "$TOTALSPINESEG"/scripts/prepare_datasets.sh [DATASET_ID] [-noaug]
+   ```
+
+   The script optionally accepts `DATASET_ID` as the first positional argument to specify the dataset to prepare. It can be either 101, 102, 103, or all. If `all` is specified, it will prepare all datasets (101, 102, 103). By default, it will prepare datasets 101 and 102.
+
+   Additionally, you can use the `-noaug` parameter to prepare the datasets without data augmentations.
```
[the remark this answers is not in the document]
Comment (yw7) [reply README.md:121]: But why anyone want to train without the data augmentation if it is not working?
Did you try without DA? Maybe you have the approximate size?

[the remark this answers is not in the document]
Comment (NathanMolinier) [reply README.md:121]: I can check yes but I did not run the full DA yet. I just think having an approximate information of the size is a good information before running the command

**Comment by NathanMolinier** [comment]
Screenshot 2024-08-29 at 09 54 45

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (yw7) [reply README.md:121]: https://github.com/neuropoly/totalspineseg/blob/1f607385f2fe46e5a6ef2788003882277a015f6f/README.md?plain=1#L89
this is with DA. can you add the info for without DA?

**Comment by NathanMolinier** [comment]
Some warning are still shown during nnUNet preprocessing but the script is still running. Are we good with it ?

Screenshot 2024-09-04 at 09 53 08

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (NathanMolinier) [reply README.md:121]: Yes you're right ! I think we are good with this point then

**Comment by NathanMolinier** [comment]
Apart from this last warning during preprocessing, everything seems to work fine ! I think we can merge !

**Comment by yw7** [comment]
> Some warning are still shown during nnUNet preprocessing but the script is still running. Are we good with it ?
> 
> 
> 
> Screenshot 2024-09-04 at 09 53 08
> 
> 

I think if the model is working we do not have to mind those.. 
Also this warning regarding diffs of 0.00000... in the affine matrix so it is not really matter ..

**Comment by yw7** [comment]
> Apart from this last warning during preprocessing, everything seems to work fine ! I think we can merge !

Amazing!!!
Thank you Nathan 🙏

**Comment by yw7** [comment]
@jcohenadad Can we merge?

**Review by NathanMolinier — APPROVED** [review]
This PR seems good to be merged for me !

**Pull request merged**